ipn/ipnlocal,net/dns/resolver: use the user dialer and routes for DNS forwarding by default, except on iOS and Android

nickkhyl opened the pull request:
[2] commits

In this PR, we make the `user-dial-routes` nodeAttr behavior default on all platforms except for iOS and Android. It can be disabled by setting the `TS_DNS_FORWARD_USE_ROUTES` envknob to `0` or `false`.

We also update `(*net/tsdial.Dialer).SetRoutes()` to log the size of the resulting `bart.Table`.

Updates #12027
Updates #13837

nickkhyl commented on net/tsdial/tsdial.go:
```diff
@@ -151,6 +151,7 @@ func (d *Dialer) SetRoutes(routes, localRoutes []netip.Prefix) {
 		for _, r := range localRoutes {
 			rt.Insert(r, false)
 		}
+		d.logf("tsdial: bart table size: %d", rt.Size())
```
`(*bart.Table).Size()` returns the number of prefixes. But in this case, it’s also a good approximation of the total size.

nickkhyl commented on net/dns/resolver/forwarder.go:
```diff
@@ -740,18 +741,36 @@ func (f *forwarder) sendUDP(ctx context.Context, fq *forwardQuery, rr resolverAn
+// we can get rid of the nodeAttr/control knob and always use UserDial for DNS.
+//
+// See https://github.com/tailscale/tailscale/issues/12027.
+func ShouldUseRoutes(knobs *controlknobs.Knobs) bool {
+	switch runtime.GOOS {
+	case "android", "ios":
+		// On mobile platforms with lower memory limits (e.g., 50MB on iOS),
+		// this behavior is still gated by the "user-dial-routes" nodeAttr.
+		return knobs != nil && knobs.UserDialUseRoutes.Load()
+	default:
+		// On all other platforms, it is the default behavior,
+		// but it can be overridden with the "TS_DNS_FORWARD_USE_ROUTES" env var.
+		return !optDNSForwardUseRoutes().EqualBool(false)
+	}
```
The `knobs` are not used on non-mobile platforms.

nickkhyl commented on net/dns/resolver/forwarder.go:
```diff
@@ -740,18 +741,36 @@ func (f *forwarder) sendUDP(ctx context.Context, fq *forwardQuery, rr resolverAn
+	default:
+		// On all other platforms, it is the default behavior,
+		// but it can be overridden with the "TS_DNS_FORWARD_USE_ROUTES" env var.
+		return !optDNSForwardUseRoutes().EqualBool(false)
```
`optDNSForwardUseRoutes().EqualBool(true)` isn’t equivalent, as it handles the unconfigured env var differently. But sure, let's make a variable.

raggi commented:
The new log output prints more often than I think we'd like: 

e.g.
```
wgengine: Reconfig: [RwLGR] changed from "discokey:8878e4f78583dc710c02f47de6a483b27d0263ae4cb9740afe738aaf02405a70" to "discokey:b631e91fffd689beb972067361cab1aeddd87534d307b861214c71b33f6d3421"
tsdial: bart table size: 4
tsdial: bart table size: 4
wgengine: Reconfig: [PpufT] changed from "discokey:de51eabc4ccb490669d6b3a309d22dac90ce88b62b1a25bbb85caee9220fcf7b" to "discokey:df7f2a14194c0f7c00fdfd9c7153f5e97ebd016b3aebcd39edca100641104232"
tsdial: bart table size: 4
tsdial: bart table size: 4
```

nickkhyl commented:
Hmm.. That's odd. It's called only called when `wgengine.Engine.Reconfig()` reports a change since the last reconfig.

But we call `(*LocalBackend).authReconfig()` in 6 different places. Notably, it may be called twice (or even thrice if the state machine transitions to `Starting` or `NeedsMachineAuth`) while handling control client status updates, and the first call uses the old netmap.

So I'm wondering if this fixes it for you: https://github.com/tailscale/tailscale/commit/4c09ea6c24a13102debb99d717949db71d40b36c
